internal/logging/sentry_replacer.go

```diff
@@ -45,13 +52,56 @@ func NewSentryReplacer(w io.Writer) *SentryReplacer {
 }
 
 func (sr *SentryReplacer) Write(p []byte) (n int, err error) {
+	if sr.closed {
+		return 0, io.EOF
+	}
+
 	sr.mu.Lock()
 	defer sr.mu.Unlock()
 
+	// Append p to our own buffer, see if there's anything we need to censor.
+	sr.buf = append(sr.buf, p...)
+
+	// If we've appended at least a line, censor it and write it out.
+	for {
+		if len(sr.buf) == 0 {
+			// Buffer flushed out completely.
+			return len(p), nil
+		}
+
+		idx := bytes.IndexRune(sr.buf, '\n')
+		if idx < 0 {
+			// No line yet, just lie to the caller and tell them we wrote p.
+			return len(p), nil
+		}
+
+		var line []byte
+		line, sr.buf = sr.buf[:idx+1], sr.buf[idx+1:]
+		line = sr.re.ReplaceAll(line, replacement)
+
+		_, err := sr.w.Write(line)
+		if err != nil {
+			// This is not strictly the error related to the incoming `p`, but the best we can do.
+			return 0, fmt.Errorf("cannot filter: %w", err)
+		}
+	}
+}
```

Comment:
So it looks like level gets replaced by the library:https://github.com/archdx/zerolog-sentry/blob/a5e7c6e3cfc1dabe917a8ce22c4a7642a0c9bafd/writer.go#L94

So we cannot delegate here, these must be two different functions.

Can you find a way to reuse this code? It looks like a complex one. How about making a helper function and then passing in an (anonymous) function that will:

* call `Write(line)`
* call `WriteLevel(level, line)` where level would be captured via closure

A callback approach is not the only one, that is the first one which comes to my mind.

Reply:
Yeah, I did think about few approaches but all felt bit dirty, I'll try something in a follow-up commit.

With callbacks I see bit of an issue of creating closure (I'm not sure how it works in Go TBH), because closures can take quite a bit of a space on stack, which might go against the minimal memory footprint of zerolog.